Addition of New Feature - Binary Support

arortiz-rh opened the pull request:
Greenboot now supports running binary files, not just script files.
- Focused changes on the function `run_scripts` in the file `greenboot.rs`. 
    - Added extra filtering to get executable files while avoiding directories. 
    - Changed command to switch between what is needed for binary files and what is needed for script files, since they require different commands to correctly run.
- Added additional tests within `greenboot.rs` and modified some existing tests slightly. 
    - This accommodates the extra files needed to test for binary files.
    - A helper function for the tests was also modified in order to set up the file structure to include the binary test files.
- Made some minor changes to the the outputs within `greenboot.rs`  to include scripts/binaries or script/binary to be more representative of what this now does.
    - Previously just said "scripts"  when communicating in outputs. 

This is my first PR and first decently sized Rust project, please make as many comments and nitpicks as you see fit.

arortiz-rh commented on rpmbuild/SPECS/greenboot-60d95276f8cbaee85a364cd47d2c81b76637a054.spec:
I don't think this file should be merged as I didn't manually create it. I believe this is something that was just automatically generated while I was creating rpms for testing.

djach7 replied:
I believe you're correct, I think this can be left out of this PR.

yih-redhat commented:
May need to add binary files under /usr/lib/greenboot/check/required.d and /usr/lib/greenboot/check/wanted.d, the binary file in required.d should success, and the binary file in wanted.d can fail, so that we can verify greenboot can run binary file as success or fail, as same as the current script files.
Please correct me if I am wrong @say-paul

say-paul commented:
Note: we will merge this after #54 , as this kind of touches the same piece of code.

yih-redhat commented:
@say-paul Do we need to provide sample binary files in /usr/lib/greenboot/check? just like existing script files.

say-paul commented:
> I also noticed that it's trying to find a directory that doesn't seem to exist, which is returning Jun 06 22:29:03 localhost greenboot[890]: grep: /etc/ostree/remotes.d/*: No such file or directory. Does anyone know what the purpose of finding this directory is for?

This healthcheck dates back to rpm-ostree days, the url from which `rpm-ostree update` will fetch the updates are stored here, which is not the case with bootc. We have a tracking story to improve this: https://issues.redhat.com/browse/THEEDGE-3108
Maybe we can move these from bash to binary once we have this PR merged. That said, this test is in the wanted section- which should not affect the greenboot flow.

arortiz-rh commented:
> > Unless I'm misunderstanding the logs, it seems to me that a fallback is detected, but the fallback log is missing.
> > TASK [assert] ****************************************************************** fatal: [192.168.100.50]: FAILED! => changed=false  assertion: '''FALLBACK BOOT DETECTED! Default bootc deployment has been rolled back'' in result_greenboot_log.stdout' evaluated_to: false msg: Fallback log not found
> 
> We can not be sure if the greenboot rolled back as the other tests related to the rollback failed too:
> 
> ```
> TASK [assert] ******************************************************************
> fatal: [192.168.100.50]: FAILED! => changed=false 
>  assertion: boot_time_log.stdout == '9'
>  evaluated_to: false
>  msg: boot times is not correct, please check configuration in /etc/greenboot/greenboot.conf

arortiz-rh commented:
> > Unless I'm misunderstanding the logs, it seems to me that a fallback is detected, but the fallback log is missing.
> > TASK [assert] ****************************************************************** fatal: [192.168.100.50]: FAILED! => changed=false  assertion: '''FALLBACK BOOT DETECTED! Default bootc deployment has been rolled back'' in result_greenboot_log.stdout' evaluated_to: false msg: Fallback log not found
> 
> We can not be sure if the greenboot rolled back as the other tests related to the rollback failed too:
> 
> ```
> TASK [assert] ******************************************************************
> fatal: [192.168.100.50]: FAILED! => changed=false 
>  assertion: boot_time_log.stdout == '9'
>  evaluated_to: false
>  msg: boot times is not correct, please check configuration in /etc/greenboot/greenboot.conf
> ```

Could the rollback not being detected be because I changed the language of the output to include "scripts/binaries" or something similar to that? I noticed it's checking `Skipping disabled script: 01_repository_dns_check.sh''` which isn't be detected, however that would no longer be the output of the program because I changed the language of the output to be `Skipping disabled script/binary: " followed by whatever script or binary or should be there. You can see this change I made at line 147 in `greenboot.rs`.

Is this something I should revert? I changed the language originally to be more reflective of what is now being checked, but if it's interfering with testing I can change it back.

djach7 commented on src/lib/greenboot.rs:
```diff
@@ -101,19 +103,35 @@ struct ScriptRunResult {
     skipped: Vec<String>,
 }
 
+// This returns a structure that contains the files within the file path that 
```
It looks like this comment got cut off?

djach7 commented on src/lib/greenboot.rs:
```diff
@@ -101,19 +103,35 @@ struct ScriptRunResult {
 
+// This returns a structure that contains the files within the file path that 
 fn run_scripts(name: &str, path: &str, disabled_scripts: Option<&[String]>) -> ScriptRunResult {
+    // Creating a new vector here to store the result of a script that's been run.
```
Is this comment in reference to `result` or `entries`? If it's `entries` then I think it can be moved down to line 113

arortiz-rh replied:
This was another old comment I made when I was first parsing through this code. I can move it, but I don't think it's necessary to leave in so I can also just remove it

djach7 commented on src/lib/greenboot.rs:
```diff
@@ -277,4 +355,5 @@ mod test {
         fs::remove_dir_all(GREENBOOT_INSTALL_PATHS[1]).expect("Unable to delete folder");
         Ok(())
     }
+
```
I think this empty line can be removed unless the linter really wants it.

djach7 reviewed (changes requested):
Getting a chance to look at this, this is a great start! I pulled it down and tested locally, everything `cargo test` related worked great for me. 

I also have some style/git feedback that's more related to how our team usually operates:
- Minimizing # of commits: We try to limit the number of commits per PR, keeping them centered around a specific fix or feature. Sarita's PR [here](https://github.com/fedora-iot/greenboot-rs/pull/54) is a good example, where every commit distinctly adds a feature, test, or fix. This helps us track history a little easier, as we can follow certain features back to specific commits. For your PR it might look like two commits, one with the main feature and then one with the tests.
- Feature branch vs main branch: This kind of falls in line with the reasoning for the commits, usually we try to work from a personal feature branch (i.e. `arortiz-rh:greenboot_rs_binary_support) as opposed to creating the PR from a personal main branch. This is good git practice and helps keep organization if you ever work on multiple features/fixes for the same repo simultaneously. 

Thanks for taking this on! If you want any help with the git stuff (like squashing, fixupping, force pushing, etc.) please don't hesitate to ping me. I'll also try to learn a bit about that testing farm issue to see if I can help there.

arortiz-rh commented:
> Getting a chance to look at this, this is a great start! I pulled it down and tested locally, everything `cargo test` related worked great for me.
> 
> I also have some style/git feedback that's more related to how our team usually operates:
> 
>     * Minimizing # of commits: We try to limit the number of commits per PR, keeping them centered around a specific fix or feature. Sarita's PR [here](https://github.com/fedora-iot/greenboot-rs/pull/54) is a good example, where every commit distinctly adds a feature, test, or fix. This helps us track history a little easier, as we can follow certain features back to specific commits. For your PR it might look like two commits, one with the main feature and then one with the tests.
> 
>     * Feature branch vs main branch: This kind of falls in line with the reasoning for the commits, usually we try to work from a personal feature branch (i.e. `arortiz-rh:greenboot_rs_binary_support) as opposed to creating the PR from a personal main branch. This is good git practice and helps keep organization if you ever work on multiple features/fixes for the same repo simultaneously.
> 
> 
> Thanks for taking this on! If you want any help with the git stuff (like squashing, fixupping, force pushing, etc.) please don't hesitate to ping me. I'll also try to learn a bit about that testing farm issue to see if I can help there.

I see, thank you for explaining this to me! I didn't even realize that the amount of commits or doing a PR from my main branch could cause some organization issues, but how you explained it makes a lot of sense. I think it might be to late for me to change this PR to a different branch, but I believe I can still squash the commits so I'll get started on trying to do that. Thank you again!

say-paul commented on src/lib/greenboot.rs:
```diff
@@ -233,11 +263,33 @@ mod test {
         let passing_test_scripts = "testing_assets/passing_script.sh";
         let failing_test_scripts = "testing_assets/failing_script.sh";
+        let passing_test_binary = "testing_assets/passing_binary";
+        let failing_test_binary = "testing_assets/failing_binary";
```
@arortiz-rh you will need to add the binaries here too: https://github.com/fedora-iot/greenboot-rs/tree/main/testing_assets

say-paul commented:
Can you rebase the PR,
also do a `cargo fmt` before pushing, we want all the checks to run green only exception is `CI Test / Fedora-43-bootc`.

yih-redhat commented:
CI job of qcow2 test on centos-10 failed because:
1. didn't find fallback log, which means fallback is probably not happened.
2. didn't find that vm rebooted 5 times before rollback, the result shows it's not rebooted.
3. didn't find health-check fail early log. (this one could be the root cause of above two failures. A failed script is installed https://github.com/fedora-iot/greenboot-rs/blob/main/tests/greenboot-bootc-qcow2.sh#L267 here, and greenboot should fail the health-check and trigger fallback, I didn't see such log there, that's why greenboot didn't fallback and reboot.)

yih-redhat commented:
Hi @arortiz-rh I added a commit (includes test case) into this pr and triggered CI test, let's wait for the test result.

Just a kind reminder (ignore it if you already know), when you push to this pr again, please rebase your local repo to get my commit, and then change and push. Otherwise my commit will be override.

yih-redhat commented:
Test case passed on qcow2-centos10, and anaconda-centos10 failed with known issue https://issues.redhat.com/browse/RHEL-96222

djach7 commented on src/lib/greenboot.rs:
```diff
@@ -213,6 +233,11 @@ mod test {
         setup_folder_structure(false)
             .context("Test setup failed")
             .unwrap();
+        // Causes errors if these are no removed since they cause an excess amount
```
Super-duper nit: `no` to `not`, `failure` to `failures`

djach7 reviewed (changes requested):
Pulled down and tested again, LGTM overall. I have one super duper grammar nit but  that's it

say-paul reviewed (changes requested):
Overall the changes looks good to me.
I would prefer the commits to scrub off all the unnecessary file changes as it will muddle the git logs and cause issue if anyone branches off from a commit.

@arortiz-rh you can merge the functionality + unit tests into one commit  and keep the integration tests(from @yih-redhat ) in a seperate commit. And should get rid of all the intermittent files added as part of 9243368596f80c9042699bb075d1a807b20e0afc which was later deleted in the next commit.

arortiz-rh commented:
> Overall the changes looks good to me. I would prefer the commits to scrub off all the unnecessary file changes as it will muddle the git logs and cause issue if anyone branches off from a commit.
> 
> @arortiz-rh you can merge the functionality + unit tests into one commit and keep the integration tests(from @yih-redhat ) in a seperate commit. And should get rid of all the intermittent files added as part of [9243368](https://github.com/fedora-iot/greenboot-rs/commit/9243368596f80c9042699bb075d1a807b20e0afc) which was later deleted in the next commit.

I merged my two commits into one and got rid of the intermittent files you mentioned. Still passing the qcow2 test for Centos, just failing the iso-anaconda due to the same problem as before. If anything else needs to be changed, please let me know!

djach7 reviewed (approved):
LGTM
